### PR TITLE
Avoid CreateFile in hot GetLastWriteFileUtcTime

### LINKED ISSUE #2052: Severe performance regression in FlatTrackingData when using CreateFileW
Originally reported at https://developercommunity.visualstudio.com/content/problem/37589/regression-debug-slow-to-start-when-always-build-i.html

Performance traces indicate that a major part of the performance problem reported comes down to calling `CreateFile()` + `GetFileTime()` instead of  `GetFileAttributesEx()` when `stat`ing files.

This change was made in #648 to supporting symlinks as source files.

<details><summary>Performance traces</summary>

```
Dev14

Name                                                                                                                                                                		  Inc
|||+ microsoft.build.utilities.core.ni![COLD] Microsoft.Build.Utilities.FlatTrackingData.ConstructFileTable()                                                      		271.818
||||+ microsoft.build.utilities.core.ni!FlatTrackingData.UpdateFileEntryDetails                                                                                    		271.313
|||||+ microsoft.build.utilities.core.ni!FlatTrackingData.RecordEntryDetails                                                                                       		264.189
||||||+ microsoft.build.utilities.core.ni!FlatTrackingData.GetLastWriteTimeUtc                                                                                     		247.695
|||||||+ microsoft.build.utilities.core.ni!NativeMethods.GetLastWriteTimeUtc                                                                                       		208.277
||||||||+ microsoft.build.utilities.core.ni!DomainBoundILStubClass.IL_STUB_PInvoke(System.String, Int32, Microsoft.Build.Utilities.WIN32_FILE_ATTRIBUTE_DATA ByRef)		207.287
|||||||||+ kernelbase!GetFileAttributesExW                                                                                                                         		205.572

Dev15

Name                                                                                                                                             		    Inc
|| + microsoft.build.utilities.core.ni![COLD] Microsoft.Build.Utilities.FlatTrackingData.ConstructFileTable()                                    		2,205.569
|| |+ microsoft.build.utilities.core!FlatTrackingData.UpdateFileEntryDetails                                                                     		2,201.341
|| ||+ microsoft.build.utilities.core.ni!FlatTrackingData.RecordEntryDetails                                                                     		2,177.880
|| |||+ microsoft.build.utilities.core.ni!FlatTrackingData.GetLastWriteTimeUtc                                                                   		2,148.838
|| ||||+ microsoft.build.utilities.core.ni!NativeMethodsShared.GetLastWriteFileUtcTime                                                           		2,055.619
|| |||||+ microsoft.build.utilities.core.ni!DomainBoundILStubClass.IL_STUB_PInvoke(System.String, UInt32, UInt32, IntPtr, UInt32, UInt32, IntPtr)		1,735.259
|| ||||||+ kernelbase!CreateFileW                                                                                                                		1,724.177
```
</details>

### PR DESCRIPTION
**Pull request by rainersigwald**
Fixes #2052 by calling the more performant `GetFileAttributesEx()`
first, and falling back to the slower `CreateFile()` + `GetFileTime()`
_only_ if `GetFileAttributesEx` reports that what it opened was a
"reparse point" (symlink).

### CODE REVIEW

**Comment by rainersigwald** [comment]
whitespace makes this confusing: https://github.com/Microsoft/msbuild/pull/2087/files?w=1 is cleaner (I nested the new codepath).

File: src/Shared/NativeMethodsShared.cs [hunk 1]
```diff
@@ -862,35 +862,37 @@ internal static MemoryStatus GetMemoryStatus()
...
                 {
-                    using (SafeFileHandle handle =
-                        CreateFile(fullPath, GENERIC_READ, FILE_SHARE_READ, IntPtr.Zero, OPEN_EXISTING,
-                            FILE_ATTRIBUTE_NORMAL, IntPtr.Zero))
+                    long dt = ((long)(data.ftLastWriteTimeHigh) << 32) | ((long)data.ftLastWriteTimeLow);
+                    fileModifiedTime = DateTime.FromFileTimeUtc(dt);
+
+                    // If file is a symlink _and_ we're not instructed to do the wrong thing, open a handle to the destination file
+                    // (no FILE_FLAG_OPEN_REPARSE_POINT) and inspect its time.
+                    if ((data.fileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) == FILE_ATTRIBUTE_REPARSE_POINT && !UseSymlinkTimeInsteadOfTargetTime)
```
Comment (jeffkl) [inline src/Shared/NativeMethodsShared.cs]: Are hard links treated any differently?

**Comment by rainersigwald** [comment]
Great point @jeffkl. The directory entry for a hard link is only updated for the link that was used to modify a file, so other hardlinks to the same content will have stale dates.

MSBuild was always wrong in that dimension, until the too-slow change that this is replacing. But I hate to take it away entirely.

What I did in my latest push is add a new environment variable `MSBUILDALWAYSCHECKCONTENTTIMESTAMP`, which will opt into the slow, accurate accounting all the time. Folks who use hardlinks can set that. Everyone else will get performant checks for regular files with pay-for-play slower but accurate checks for symlinks.

@Microsoft/msbuild-maintainers does this sound like a reasonable compromise?

**Comment by rainersigwald** [comment]
From standup: use the new feature-flags/escape-hatch class instead of direct environment check.

File: src/Build.OM.UnitTests/Microsoft.Build.Engine.OM.UnitTests.csproj [hunk 2]
```diff
@@ -55,6 +55,9 @@
     <Compile Include="..\Shared\ResourceUtilities.cs">
       <ExcludeFromStyleCop>true</ExcludeFromStyleCop>
     </Compile>
+    <Compile Include="..\Shared\Traits.cs">
+      <Link>Traits</Link>
```
Comment (KirillOsenkov) [inline src/Build.OM.UnitTests/Microsoft.Build.Engine.OM.UnitTests.csproj]: Does this mean it goes in the Traits subfolder or is the .cs extension missing?

Comment (rainersigwald) [inline src/Build.OM.UnitTests/Microsoft.Build.Engine.OM.UnitTests.csproj]: I don't know, but it makes much more sense to call it `SharedTraits.cs` so I did. Nice catch.

Comment (cdmihai) [inline src/Build.OM.UnitTests/Microsoft.Build.Engine.OM.UnitTests.csproj]: Does this make it look like it's SharedTraits.cs inside VS? That might be confusing :)

---
In reply to: [116095729](https://github.com/Microsoft/msbuild/pull/2087#discussion_r116095729) [](ancestors = 116095729)

**Comment by AndyGerlicher** [comment]
@dotnet-bot test Windows_NT Build for Desktop please

**Comment by cdmihai** [comment]
@dotnet-bot test Windows_NT Build for Desktop please

Flaky restore

**Comment by rainersigwald** [comment]
CI is a mystery.

**Pull request merged**